app/controllers/oauth2/authorization_controller.rb

```diff
@@ -0,0 +1,150 @@
+
+    @app = client
+    # If it's already authorized, just ask for the new scopes
+    @scopes = authorization ? scopes - authorization.scopes : scopes
+    @previous_scopes = authorization ? authorization.scopes : []
+    @redirect_uri = params[:redirect_uri]
+    @state = params[:state] if params.key?(:state)
+
+    return fatal_error! :invalid_client unless client
+    return fatal_error! :readonly_client unless client.write_access?
+    return fatal_error! :invalid_redirect unless valid_redirect?
+    return error! :unsupported_response_type unless valid_response_type?
+    return error! :invalid_scope unless valid_scopes?
+
+    render :ask
+  rescue
+    if Rails.env.production?
+      error! :server_error
+    else
+      raise
+    end
+  end
+
+  # User confirmed that they wish to authorize the client, redirect back to
+  # client.
+  def give
+    # If yes, issue OAuth2::Code and redirect to client
+    if params[:accept]
+      auth = authorization || ClientAuthorization.new(user: current_user,
+                                                      app: client)
+      auth.scopes = scopes
+      auth.save!
+      redirect_back_with_token!
+    else
+      return error! :access_denied
+    end
+  rescue
+    if Rails.env.production?
+      error! :server_error
+    else
+      raise
+    end
+  end
+
+  private
+
+  def scopes
+    params[:scope] ? params[:scope].split(' ') : []
+  end
+
+  def client
+    @app ||= App.where(key: params[:client_id]).first
+    @app
+  end
+
+  def authorization
+    @authorization ||= ClientAuthorization.where(user: current_user,
+                                                 app: client)
+                                          .first
+    @authorization
+  end
+
+  def valid_redirect?
+    client.redirect_allowed?(params[:redirect_uri])
+  end
+
+  def valid_scopes?
+    client.scopes_allowed?(scopes)
+  end
+
+  def valid_response_type?
+    %w[code token].include?(params[:response_type])
+  end
+
+  # OAuth2 has specific error protocols to report back to the redirect endpoint
+  # since normal HTTP status codes don't work (4.1.2.1, 4.2.2.1)
+  def error!(err, message=nil, uri: nil)
+    error! :server_error unless VALID_ERRORS.include?(err)
+    redirect_back!({
+      error: err,
+      error_description: message,
+      error_uri: uri
+    }.compact!)
+  end
+
+  # Fatal errors are ones which should be displayed to the user without calling
+  # the redirect endpoint. ex: invalid client_id or redirect_uri (4.1.2.1P1)
+  def fatal_error!(err)
+    @error = err
+    render :error, status: 400
+  end
+
+  # Redirect back to the application with payload
+  def redirect_back!(obj)
+    # State parameter to help clients prevent CSRF (10.12)
+    obj = { state: params[:state] }.merge(obj).compact
+
+    uri = if params[:response_type] == 'token'
+            client.build_redirect_uri(:fragment, obj)
+          else
+            client.build_redirect_uri(:query_string, obj)
+          end
```

Comment:
Won't be better to read to use a ternary rather than an if on the assignment?
```ruby
type = params[:response_type] == 'token' ? :fragment : :query_string
uri = client.build_redirect_uri(type, obj)